File: 6/6.py
def detect_marker(signal: str, length) -> int:
    split = list(signal)
    buffer = []
    counter = 0
    while split:
        while len(buffer) < length:
            buffer.append(split.pop(0))
            counter += 1
        if len(buffer) == len(set(buffer)):
            return counter
        else:
            buffer.pop(0)
    raise ValueError


def part_1() -> None:
    tests = (
        ("mjqjpqmgbljsphdztnvjfqwrcgsmlb", 7),
        ("bvwbjplbgvbhsrlpgdmjqwftvncz", 5),
        ("nppdvjthqldpwncqszvftbrmjlhg", 6),
        ("nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg", 10),
        ("zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw", 11),
    )
    for raw, expected in tests:
        assert detect_marker(raw, 4) == expected
    with open("6_input.txt", "r") as f:
        data = f.read()
        print(detect_marker(data, 4))


def part_2() -> None:
    tests = (
        ("mjqjpqmgbljsphdztnvjfqwrcgsmlb", 19),
        ("bvwbjplbgvbhsrlpgdmjqwftvncz", 23),
        ("nppdvjthqldpwncqszvftbrmjlhg", 23),
        ("nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg", 29),
        ("zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw", 26),
    )
    for raw, expected in tests:
        assert detect_marker(raw, 14) == expected
    with open("6_input.txt", "r") as f:
        data = f.read()
        print(detect_marker(data, 14))


def main():
    part_1()
    part_2()


main()

File: 6/6.cc
// #include <iostream>
#include <fstream>
#include <iostream>
#include <string.h>
#include <map>
#include <err.h>

int detect_marker(std::string input, int length)
{
    std::string buffer;
    int counter = 0;
    while (input.size() > 0)
    {
        while (buffer.size() < length)
        {
            buffer.push_back(input[0]);
            input.erase(0, 1);
            ++counter;
        }
        std::map<char, int> lookup;
        for (int i = 0; i < buffer.size(); ++i)
        {
            lookup[buffer[i]]++;
        }
        if (lookup.size() == length)
        {
            return counter;
        }
        else
        {
            buffer.erase(0, 1);
        }
    }
    throw std::range_error("Run out of letters");
}

void part_1_tests()
{
    assert(detect_marker("mjqjpqmgbljsphdztnvjfqwrcgsmlb", 4) == 7);
    assert(detect_marker("bvwbjplbgvbhsrlpgdmjqwftvncz", 4) == 5);
    assert(detect_marker("nppdvjthqldpwncqszvftbrmjlhg", 4) == 6);
    assert(detect_marker("nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg", 4) == 10);
    assert(detect_marker("zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw", 4) == 11);
}

void part_2_tests()
{
    assert(detect_marker("mjqjpqmgbljsphdztnvjfqwrcgsmlb", 14) == 19);
    assert(detect_marker("bvwbjplbgvbhsrlpgdmjqwftvncz", 14) == 23);
    assert(detect_marker("nppdvjthqldpwncqszvftbrmjlhg", 14) == 23);
    assert(detect_marker("nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg", 14) == 29);
    assert(detect_marker("zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw", 14) == 26);
}

int main()
{
    // run all tests for part 1 first
    part_1_tests();

    // Run on input file
    std::ifstream file("6_input.txt");
    std::string buffer;
    while (getline(file, buffer))
    {
        int result = detect_marker(buffer, 4);
        std::cout << result << std::endl;
    }

    // run all tests for part 1 first
    part_2_tests();

    // Run on input file
    std::ifstream file2("6_input.txt");
    while (getline(file2, buffer))
    {
        int result = detect_marker(buffer, 14);
        std::cout << result << std::endl;
    }

    return 0;
}